plutus-core-spec/plutus-core-specification.tex

```diff
@@ -1033,6 +1027,99 @@ \section{Examples}
+Figure~\ref{fig:untyped-values}.
+
+%% The introduction of \texttt{delay} and \texttt{force} may be
+%% controversial since $\delay{M}$ can be implemented as $\lamU{u}{M}$
+%% and $\force{M}$ as $\appU{M}{\lamU{x}{x}}$.  However, doing this
+%% has a cost in code size.  Experiments show that
+%% adding \texttt{delay} and \texttt{force} increases the size of
+%% compressed validators by about 10\%, but using $\lamU{u}{M}$ and
+%% $\appU{M}{\lamU{x}{x}}$ increases sizes by about 20\%.  A third
+%% option is to introduce a global variable $u$ of type \texttt{unit}
+%% and replace $\force{M}$ with $\appU{M}{u}$: this only increases
+%% sizes by about 12\%, but complicates specification because programs
+%% have to be wrapped in an expression binding $u$ (and maybe the CEK
+%% machine has to start with an environment binding $u$ to
+%% $\lamU{x}{x}$).  Another option would be to extend the grammar with
+%% a unit term, but then it would be tempting to add terms
+%% for \texttt{true}, \texttt{false}, and so on.  It's possible that
+%% in actual validator code (especially after type
+%% normalisation), \texttt{delay} and \texttt{force} will typically
+%% appear in expressions of the form $\force{\delay{e}}$, which could
+%% immediately be replaced by $e$ alone, and then using $\lamU{x}{x}$
+%% might not cause so much overhead. I'll look into this.  Later:
+%% nope.  I looked at the on-chain validators for the use cases and
+%% this only happens about 10\% of the time.
+
+\subsection{Type erasure}
+Figure~\ref{fig:type-erasure} defines a type erasure operation which
+converts a typed Plutus Core term $M$ into an untyped one
+$\erase{M}$ (objects from the typed grammar on the left of $=$,
+untyped on the right).  %\footnotemark[4].
+We extend this to sequences of
+terms in the obvious way: given $M^* = M_1 \ldots M_n$ we define
+$\erase{M^*} = \erase{M_1} \ldots \erase{M_n}$.  For a typed program
+$P$ we use $\erase{P}$ to denote the untyped program with the same
+version whose body is the erasure of the body of $P$. The erasure operation
+preserves semantics in an appropriate sense.  It can be shown that if $P$
+is a well-typed Plutus Core program then $P$ terminates if 
+and only if $\erase{P}$ does, and that $P$ returns the value $v$
+if and only if $\erase{P}$ returns the value $\erase{v}$.
+
+%% I'm not wedded to this particular notation for erasure,
+%% but I think it's unlikely that it'll confuse people. Other
+%% possiblities: $\left|M\right|$, $\left\lVert{M}\right\rVert$,
+%% $\overline{M}$ (that looks pretty bad), etc. I'll consult some experts to
+%% find out if there's a standard notation.  (Later: apparently not.)
+
+\subfile{figures/TypeErasure}
+
+\subsection{Term Reduction}
+Figure~\ref{fig:untyped-term-reduction} defines a reduction semantics
+for untyped Plutus Core, analogously to
+Figure~\ref{fig:term-reduction}.  The built-in functions in
+Figure~\ref{fig:builtins} are interpreted in the untyped setting in the
+obvious way.
```

Comment:
I'm not sure that this is obvious! TBH, the semantics of builtins in the spec are underspecified, e.g. Fig 15 gives semantics in terms of integers for integer operations, but doesn't specify what *term* that produces. Probably we need some rule for builtins like:
- A builtin can only reduce to:
    - A `con`
    - One if its arguments (to handle polymorphic builtins)

Then I think we'd be okay with erasure.

Reply:
Yeah, I fudged that a bit; I'll think about that a bit more.  We'll have to deal with extensible builtins at some point, and that may be a bit tricky. I think we'll need two explanations: a technical one saying how things work in the Haskell implementation (probably a separate document), and some more general thing in this document saying that you're free to extend the builtins however your evaluator is implemented (and no matter what language you're using), but you have to satisfy some yet-to-be-determined restrictions.  For example, presumably polymorphic builtins have to be parametrically polymorphic, to avoid having dynamically-typed Plutus Core, which would surely be a bad thing.

Reply:
Having slept on this, I've decided that it's a real nuisance.  Without the types, there's nothing to stop us applying builtins to nonsensical arguments, and I don't think there's any way to overcome this without placing a lot of conditions on how the builtins behave.  I've made an attempt at dealing with this by saying that the untyped semantics are defined only for terms which we've got by erasing typed ones and then lifting the behaviour of builtins on untyped terms back to the typed semantics.  I think this means that you really do have to confirm off-chain that your untyped validator is obtained from a well-typed one (which you actually have to be in possession of, because type reconstruction is undecidable) . If you allow arbitrary untyped code to run on-chain and your evaluator happens to be implemented in C then all bets are off if a builtin is fed get arguments of the wrong sort.